z3c/dependencychecker/db.py

```diff
@@ -39,7 +44,234 @@ def add_extra_requirements(self, extra_name, dotted_names):
+            self._filter_out_python_standard_library,
+        )
+        missing = self._process_pipeline(self.imports_used, filters)
+        unique_imports = self._get_unique_imports(imports_list=missing)
+        return unique_imports
+
+    def get_unneeded_requirements(self):
+        all_but_test_requirements = self._requirements.copy()
+        for extra in self._extras_requirements:
+            all_but_test_requirements.update(self._extras_requirements[extra])
+        test_requirements = self._get_test_extra()
+        if test_requirements:
+            for dotted_name in test_requirements:
+                all_but_test_requirements.remove(dotted_name)
+        filters = (
+            self._filter_out_known_packages,
+            self._filter_out_python_standard_library,
+            self._filter_out_used_imports,
+        )
+        unneeded = self._process_pipeline(all_but_test_requirements, filters)
+        unique_imports = self._get_unique_imports(imports_list=unneeded)
+        return unique_imports
+
+    def requirements_that_should_be_test_requirements(self):
+        non_testing_filters = (
+            self._filter_out_testing_imports,
+            self._filter_out_python_standard_library,
+            self._filter_out_own_package,
+        )
+        non_testing_imports = self._process_pipeline(
+            self.imports_used,
+            non_testing_filters,
+        )
+        requirements_not_used = [
+            requirement
+            for requirement in self._requirements
+            if self._discard_if_found_obj_in_list(
+                requirement,
+                non_testing_imports,
+            )
+        ]
+        testing_filters = (
+            self._filter_out_only_testing_imports,
+            self._filter_out_python_standard_library,
+            self._filter_out_own_package,
+        )
+        testing_imports = self._process_pipeline(
+            self.imports_used,
+            testing_filters,
+        )
+        should_be_test_requirements = [
+            requirement
+            for requirement in requirements_not_used
+            if not self._discard_if_found_obj_in_list(
+                requirement,
+                testing_imports,
+            )
+        ]
+        unique = self._get_unique_imports(
+            imports_list=should_be_test_requirements,
+        )
+        return unique
+
+    def get_unneeded_test_requirements(self):
+        test_requirements = self._get_test_extra()
+        if not test_requirements:
+            return []
+
+        filters = (
+            self._filter_out_known_packages,
+            self._filter_out_python_standard_library,
+            self._filter_out_used_imports,
+        )
+        unneeded = self._process_pipeline(test_requirements, filters)
+        unique_imports = self._get_unique_imports(imports_list=unneeded)
+        return unique_imports
+
+    @staticmethod
+    def _process_pipeline(objects, filters):
```

Comment:
I keep finding it hard to read this part of the code. I've been thinking about it for more than an hour now, trying to come up with a small naming change or so that might clear it up. But I just cannot figure out nice clear names.

I think there are two possible reasons:

a) Double or triple negatives. "filter_out" is negative. "if not" is negative. "discard from" is negative. "if not filtered out" is a double negative. "if not filtered out.... else..." is *another* step. Double negatives are hard to reason about (and thus it is easy to make a mistake with them).

b) I think you're trying to hard. Breaking loops early to make it efficient. Splitting everything into small functions. The `discard if obj found in list` re-implementation of basic comparison functions to make it more efficient. Aren't you re-implementing python's build-in and very efficient `set` operators?

Is it really necessary? There won't be so many imports that stuff won't fit into a list, I guess. And if you're going to go through every list/generator multiple times, you can just turn it into a list. Or a set. Most of the `_process_pipelines()`-using functions can be rewritten something like this:

    test_requirements = self._get_test_extra()
    unneeded = set(test_requirements) - set(known_packages) - set(standard_lib) - ...
    return sorted(unique_imports)

At least, that's what it looks like to me. It is also much more explicit. A "bit" more brute-force. But I don't mind if running dependencychecker takes 5 seconds or 2 seconds :-)

OTOH:

- Perhaps it is needed for what comes in the last PR.
- The code works, we can always refactor it later.

But: I'm a bit afraid of the complexity.

Reply:
I get your point and I see that I totally missed the ``set`` builtin, shame on me, but at this late on the game, I would rather prefer to keep it as it is (I can rename it to ``apply_filters`` of course) and refactor it later to use sets, would that be ok with you?

The more it goes, the more I'm interested in getting a version out in pypi and python 3 compatible 🙂

Reply:
Valid point :)
Then we'll refactor it to sets afterwards. We can fix the method name then, too. I'll just merge the pull request :)